File: java/src/main/java/com/example/cote/GJM/미나스아노르.java
package com.example.cote.GJM;

import static java.util.stream.Collectors.*;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.stream.*;

public class 미나스아노르 {

  private static final int N = 100;

  private static final double PI = Math.PI;
  private static final int INF = 987654321;

  private static final double[] x = new double[N];
  private static final double[] y = new double[N];
  private static final double[] r = new double[N];

  private static List<Range> ranges;

  private static class Range {
    Double begin;
    Double end;

    public Range(double begin, double end) {
      this.begin = begin;
      this.end = end;
    }
  }

  private static void convertToRanges(int n) {
    ranges = new LinkedList<Range>();

    for (int i = 0; i < n; i++) {
      double loc = (Math.atan2(y[i], x[i]) + 2 * PI) % (2 * PI);
      double range = 2 * Math.asin(r[i] / 2.0 / 8.0);
      ranges.add(new Range(loc - range, loc + range));
    }
  }

  private static int solveLinear(double begin, double end, int n) {

    int count = 0;
    int i = 0;

    while (begin < end) {
      double maxCover = -1;

      while (i < n && ranges.get(i).begin <= begin) {
        maxCover = Math.max(maxCover, ranges.get(i).end);
        i++;
      }

      if (maxCover <= begin) return INF;

      begin = maxCover;
      count++;
    }

    return count;
  }

  private static int solveCircular(int n) {
    int ret = INF;

    Collections.sort(
        ranges,
        (Range r1, Range r2) -> {
          if (r1.begin == r2.begin) return r1.end.compareTo(r2.end);
          return r1.begin.compareTo(r2.begin);
        });

    for (Range r : ranges) {
      System.out.println(r.begin + " , " + r.end);
    }

    for (Range r : ranges) {
      if (r.begin <= 0 || r.end >= 2 * PI) {
        double begin = r.end % (2 * PI);
        double end = (r.begin + 2 * PI) % (2 * PI);
        ret = Math.min(ret, 1 + solveLinear(begin, end, n));
      }
    }

    return ret;
  }

  public static String solve(int n) {

    convertToRanges(n);

    int ans = solveCircular(n);

    return (ans >= INF) ? "IMPOSSIBLE" : "" + ans;
  }

  public static void main(String[] args) throws IOException {

    Path p = Paths.get(System.getProperty("user.dir") + "/data/미나스아노르.txt");
    BufferedReader reader = Files.newBufferedReader(p);

    int test_cases = Integer.parseInt(reader.readLine());

    while (test_cases > 0) {

      int n = Integer.parseInt(reader.readLine());

      for (int i = 0; i < n; i++) {
        List<Double> v =
            Arrays.stream(reader.readLine().split(" ")).map(Double::parseDouble).collect(toList());

        x[i] = v.get(0);
        y[i] = v.get(1);
        r[i] = v.get(2);
      }

      System.out.println(solve(n));

      test_cases--;
    }
  }
}

File: c++/cote/GJM/미나스아노르.cpp
#include <cmath>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#define C 50
#define N 100
#define INF 987654321

using namespace std;

const double pi = 2.0 * acos(0);

double x[N], y[N];
double r[N];

vector<double>& split(const string& s) {
  stringstream ss(s);
  string buf;

  vector<double>* ret = new vector<double>();

  while (getline(ss, buf, ' ')) {
    ret->push_back(stod(buf));
  }

  return *ret;
}

vector<pair<double, double>>& convetToRange(int n) {
  vector<pair<double, double>>* ret = new vector<pair<double, double>>();

  for (int i = 0; i < n; i++) {
    double loc = fmod(2 * pi + atan2(y[i], x[i]), 2 * pi);
    double range = 2 * asin(r[i] / 2.0 / 8.0);
    ret->push_back(make_pair(loc - range, loc + range));
  }

  return *ret;
}

int solveLinear(double begin, double end, vector<pair<double, double>>& ranges,
                int n) {
  int used = 0;
  int i = 0;

  while (begin < end) {
    double maxCover = -1;

    while (i < n & ranges[i].first <= begin) {
      maxCover = max(maxCover, ranges[i].second);
      i++;
    }
    if (maxCover <= begin) return INF;
    begin = maxCover;
    ++used;
  }

  return used;
}

int solveCircular(vector<pair<double, double>>& ranges, int n) {
  int ret = INF;
  sort(begin(ranges), end(ranges));

  for (int i = 0; i < n; i++) {
    if (ranges[i].first <= 0 || ranges[i].second >= 2 * pi) {
      double begin = fmod(ranges[i].second, 2 * pi);
      double end = fmod(ranges[i].first + 2 * pi, 2 * pi);
      ret = min(ret, 1 + solveLinear(begin, end, ranges, n));
    }
  }

  return ret;
}

string solve(int n) {
  vector<pair<double, double>> ranges = convetToRange(n);

  int ret = solveCircular(ranges, n);

  return (ret >= INF) ? "IMPOSSIBLE" : to_string(ret);
}

int main(void) {
  fstream fd("data/미나스아노르.txt");
  string buf;

  getline(fd, buf);
  int test_cases = stoi(buf);

  while (test_cases > 0) {
    getline(fd, buf);
    int n = stoi(buf);
    int cur = 0;

    while (cur < n) {
      getline(fd, buf);
      vector<double> v = split(buf);

      x[cur] = v[0];
      y[cur] = v[1];
      r[cur] = v[2];
      cur++;
    }

    cout << solve(n) << endl;

    test_cases--;
  }

  return 0;
}